addon/-private/resources.ts

```diff
@@ -0,0 +1,133 @@
+import {
+  setHelperManager,
+  capabilities as helperCapabilities,
+  TemplateArgs,
+  HelperManager,
+  HelperDefinition,
+} from '@ember/helper';
+import {
+  createCache,
+  getValue,
+  Cache,
+} from '@glimmer/tracking/primitives/cache';
+import { setOwner } from '@ember/application';
+import {
+  destroy,
+  registerDestructor,
+  associateDestroyableChild,
+} from '@ember/destroyable';
+
+type Owner = unknown;
+
+type Thunk = (...args: any[]) => void;
+
+export abstract class Resource<
+  Args extends TemplateArgs = TemplateArgs,
+  T = unknown
+> {
+  protected readonly args!: Args;
+
+  constructor(ownerOrThunk: Owner | Thunk, args: Args) {
```

Comment:
In order to allow the following call without type errors, the `args` needs to be optional:

``` js
  @use user = new FetchData(() => [
    `www.example.com/users/${this.args.userId}`
  ]);
```

```suggestion
  constructor(ownerOrThunk: Owner | Thunk, arg?s: Args) {
```